support/ci/deploy.sh

```diff
@@ -48,33 +47,63 @@ fi
 
 rm ./core.sig.key
 
-COMPONENTS=($COMPONENTS)
-for component in "${COMPONENTS[@]}"
+PACKAGES=($PACKAGES)
+for package in "${PACKAGES[@]}"
 do
-  echo "--> Clearing any pre-exisiting $HAB_ORIGIN secret keys from the Studio"
-  env HAB_ORIGIN= ${TRAVIS_HAB} studio run sh -c \'rm -f /hab/cache/keys/*-*.sig.key\'
-
-  echo "--> Building $component"
-  ${TRAVIS_HAB} studio run HAB_CARGO_TARGET_DIR=/src/target build components/${component}
-
-  source ./results/last_build.env
-  HART="./results/$pkg_artifact"
-  ${TRAVIS_HAB} pkg install $HART
-
-  # once we have built the stuio, switch over to bits built here
-  if [[ "${component}" == "studio" ]]; then
-    TRAVIS_HAB=$(find /hab/pkgs/core/hab -type f -name hab | tail -1)
-  elif [[ "${component}" == "hab" ]]; then
-    RELEASE="${HART}_keep"
-    cp $HART $RELEASE
+  # Always build the package if it's a release; otherwise, only build a
+  # package if there are relevant code changes.
+  if [ -n "${IS_RELEASE}" ] || ./support/ci/should_we_build.sh "${package}" ; then
+
+     echo "Building ${package}..."
+
+     echo "--> Clearing any pre-exisiting $HAB_ORIGIN secret keys from the Studio"
+     env HAB_ORIGIN= ${TRAVIS_HAB} studio run sh -c \'rm -f /hab/cache/keys/*-*.sig.key\'
+
+     # The names of the Habitat packages we create are often different
+     # from the directories in which they reside; specifically, we
+     # often add the prefix "hab-" to the name of the package, but
+     # leave this off of the directory for ease of navigation. Here,
+     # we normalize this.
+     if [ -d "components/${package}" ]; then
+         directory_name=${package}
+     else
+         truncated_name=${package##hab-}
+         if [ -d "components/${truncated_name}" ] ; then
+             directory_name=${truncated_name}
+         else
+             echo "Cannot find code directory for package '${package}' after looking in the following locations:"
+             echo
+             echo "    components/${package}"
+             echo "    components/${truncated_name}"
+             echo
+             exit 1
+         fi
+     fi
+
+     ${TRAVIS_HAB} studio run HAB_CARGO_TARGET_DIR=/src/target build components/${directory_name}
+
+     source ./results/last_build.env
+     HART="./results/$pkg_artifact"
+     ${TRAVIS_HAB} pkg install $HART
+
+     # once we have built the stuio, switch over to bits built here
+     if [[ "${package}" == "hab-studio" ]]; then
+         TRAVIS_HAB=$(find /hab/pkgs/core/hab -type f -name hab | tail -1)
+     elif [[ "${package}" == "hab" ]]; then
+         RELEASE="${HART}_keep"
```

Comment:
if `hab` is not chosen to be built then `RELEASE` will not be set and publishing to bintray will fail. We might want to install `hab` from the unstable channel and default `RELEASE` to its hab hart in the artifact cache. Also, if we do not build the `bintray_publish` component I think the `hab pkg exec` on `hab-bintray-publish` will fail so we might want to `hab pkg install` that too.

Reply:
Hrm... good catch! I'd been thinking more about the release scenario, in which case `hab` would always be built, but this could definitely cause issues on garden-variety `master` branch builds.